base/intfuncs.jl

```diff
@@ -200,6 +200,38 @@ end
 # However, we still need a fallback that calls the general ^:
 ^{p}(x, ::Type{Val{p}}) = x^p
 
+# for numbers, inline the power_by_squaring method for literal p.
+# (this also allows integer^-p to give a floating-point result
+#  in a type-stable fashion).
+@generated function ^{p}(x::Number, ::Type{Val{p}})
```

Comment:
This really shouldn't be `@generated`. It's too likely to cause runtime performance issues with limited type information and potentially other issues with compile-time convergence / termination.

The `power_by_squaring` definition here is also inaccurate except for `::Integer` (llvm already implements this optimization for those), and `Complex{<:Integer}`. This definition would likely also be slower for non-bitstypes, such as `BigInt`.